### DISCUSSION TITLE
How to use Viper with Cobra?

### DISCUSSION THREAD

**Discussion by sagikazarmark** [spf13/viper#1061] [Q&A] [answered]
There is a lot of confusion around using Viper and Cobra together and a lot of issues seem to land in the Viper repo, but this is actually more of an integration question on the Cobra side.

This thread is supposed to provide some clues and should be a goto source for users.

**Comment by sagikazarmark** [comment, marked as answer]
Historically both Viper and Cobra originate from Hugo, a static page generator, but the individual libraries don't have any tight integration between them. This might change in the future though: with Viper 2, consumers (like Cobra) will have much more control over how they integrate with Viper.

Nevertheless, it'll still be Cobra's job to handle the integration, not the other way around, so this probably won't be "fixed" on the Viper side.

Most of the issues are around handling flags. Internally, Cobra allows defining multiple FlagSets whereas usually there is only one Viper instance. This is a problem, when multiple FlagSets define the same flags.

Another common problem is Cobra's `Required` flag feature is not supported by Viper (at least by other config sources than flags). This is also something that won't be fixed on Viper's side.

Some workarounds for :

- a [script](https://github.com/spf13/viper/issues/82#issuecomment-403165788) for copying flags (thanks to @WGH-)
- an [integration package](https://github.com/spf13/viper/issues/82#issuecomment-475910342) (thanks to @desdic)
- another [integration package](https://github.com/spf13/viper/issues/82#issuecomment-753555375) (thanks to @ cpliakas)

Feel free to share more snippets below!

**Reply by justinfx** [reply]
Following up here from https://github.com/spf13/viper/issues/461, I have found it confusing when using Cobra + Viper and the expected behaviour of binding viper to cobra flags. My expectation is that "bind" implies bi-directional, but it would seem that it only means flag values that are directly set will propagate to viper and not the other way around. This ends up meaning that if a project uses viper and binding, then it must only go through viper for all configuration values and not cobra.

Example from #461 describes the following steps:

1. define a flag with cobra
2. bind viper to cobra flag
3. bind viper to env var
4. run application with env var set, and flag unset
5. observe that the env var value is not accessible through the cobra flag api
6. observe that the env var value is accessible through the viper api
